src/ck-ldb/TreeLB.C

```diff
@@ -0,0 +1,549 @@
+{
+#if CMK_LBDB_ON
+  startTime = CkWallTimer();
+  if (CkMyPe() == 0 && _lb_args.debug() > 0)
+  {
+    CkPrintf("--------- Started LB step %d ---------\n", lbmgr->step());
+  }
+  // CmiAssert(CmiNodeAlive(CkMyPe()));   // TODO move this logic to LBManager
+  int level = 0;  // load balancing starts at the lowest level
+  CkAssert(numLevels > 0 && !awaitingLB[level]);
+  TreeLBMessage* stats = logic[level]->getStats();
+  stats->level = level;
+  awaitingLB[level] = true;
+  sendStatsUp((CkMessage*)stats);
+#endif
+}
+
+// send stats up using the comm-tree for this level
+void TreeLB::sendStatsUp(CkMessage* msg)
+{
+  TreeLBMessage* stats = (TreeLBMessage*)msg;
+  int level = stats->level;
+  int comm_parent_pe = comm_parent[level];
+  // fprintf(stderr, "[%d] TreeLB::sendStatsUp - received msg level=%d comm_parent=%d\n",
+  // CkMyPe(), level, comm_parent_pe);
+  if (comm_parent_pe == -1)
+  {
+    // I'm the root of this comm-tree (current destination of the stats)
+    receiveStats(stats, level);
+  }
+  else if (comm_children[level].size() == 0)
+  {
+    // don't have children so don't have to aggregate any msgs
+    thisProxy[comm_parent_pe].sendStatsUp((CkMessage*)stats);
+  }
+  else
+  {
+    LevelLogic* logic = comm_logic[level];
+    logic->depositStats(stats);
+    if (logic->numStatsReceived() == comm_children[level].size() + 1)
+    {
+      TreeLBMessage* newMsg = logic->mergeStats();
+      newMsg->level = level;
+      thisProxy[comm_parent_pe].sendStatsUp((CkMessage*)newMsg);
+    }
+  }
+}
+
+void TreeLB::receiveStats(TreeLBMessage* stats, int level)
+{
+  level += 1;
+  awaitingLB[level] = true;
+  LevelLogic* l = logic[level];
+  l->depositStats(stats);
+  size_t expected_msgs = comm_children[level - 1].size();
+  if (logic[level - 1] != nullptr) expected_msgs += 1;  // expect msg from myself too
+  if (l->numStatsReceived() == expected_msgs)
+  {
+    CkAssert(load_sent[level] == 0 && load_received[level] == 0);
+    if (level == numLevels - 1 || l->cutoff())
+    {
+      // cutoff can be adjusted dynamically, to prevent lb between upper-level domains.
+      // can be used, for example, to only do within-node lb on some steps
+      loadBalanceSubtree(level);
+    }
+    else
+    {
+      TreeLBMessage* newMsg = l->mergeStats();
+      newMsg->level = level;
+      sendStatsUp((CkMessage*)newMsg);
+    }
+  }
+}
+
+void TreeLB::loadBalanceSubtree(int level)
+{
+  if (!awaitingLB[level]) return;
+  awaitingLB[level] = false;
+  if (level == 0) return lb_done();
+
+  // CkPrintf("[%d] TreeLB::loadBalanceSubtree - level=%d\n", CkMyPe(), level);
+
+  std::vector<TreeLBMessage*> decisions;
+  /// CkMessage *inter_subtree_migrations = nullptr;
+  IDM idm;
+  logic[level]->loadBalance(decisions, idm);
+  if (idm.size() > 0)
+  {
+    // this can happen when final destinations of chares has been decided,
+    // and chares from a subtree need to migrate to a PE in a different subtree
+    std::vector<int> idm_dests;
```

Comment:
It would be better to construct the vector to size, then assign directly to it in the for loop